QubitServer/src/main/scala/org/labrad/qubits/QubitContext.scala

```diff
@@ -878,24 +878,39 @@ class QubitContext(cxn: ServerConnection, resources: () => Resources) extends Se
       fpga.checkTimerStatus();
     }*/
 
-    // check microwave source configuration
+    // Check microwave source configuration.
+    //
+    // We group the IQ channels by the microwave source they use and check that
+    // all channels sharing a particular source agree on its configuration. For
+    // any microwave boards which are not being used by IQ channels, we
+    // configure their sources to be off.
+
     val uwaveConfigs = mutable.Map.empty[MicrowaveSource, MicrowaveSourceConfig]
-    for (ch <- expt.getChannels[IqChannel]) {
-      val src = ch.getMicrowaveSource()
-      val config = ch.getMicrowaveConfig()
-      require(config != null, s"No microwaves configured for channel '${ch.name}'")
-      uwaveConfigs.get(src) match {
-        case None =>
-          // keep track of which microwave sources we have seen
-          uwaveConfigs(src) = config
-
-        case Some(existing) =>
-          // check that microwave configurations are compatible
-          require(config == existing,
-              s"Conflicting microwave configurations for source '${src.name}'")
-      }
+
+    val configsBySrc = (for {
+      dev <- expt.devices
+      chan <- dev.getChannels[IqChannel]
+    } yield {
+      val src = chan.getMicrowaveSource()
+      val config = chan.getMicrowaveConfig()
+      require(config != null, s"No microwaves configured: dev=${dev.name}, channel=${chan.name}")
+      src -> (dev, chan, config)
+    }).groupBy { case (src, chans) => src }
+      .mapValues(_.map { case (src, info) => info })
```

Comment:
This threw me for a loop because (1) I wasn't familiar with `groupBy` and (2) it calls the triple `(dev, chan, config)` `chans` in the groupBy and `info` in the mapValues so it wasn't obvious they were the same thing.